### PR TITLE
RegularGrid as a Mesh

### PR DESCRIPTION
**Pull request by BotellaA**

### CODE REVIEW

File: include/geode/mesh/core/geode_regular_grid_solid.h [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+        void set_vertex( index_t vertex_id, Point3D point, OGRegularGridKey );
+
+        void set_polyhedron_vertex( const PolyhedronVertex& polyhedron_vertex,
+            index_t vertex_id,
+            OGRegularGridKey );
+
+        void set_polyhedron_adjacent( const PolyhedronFacet& polyhedron_facet,
+            index_t adjacent_id,
+            OGRegularGridKey );
```
Comment (BotellaA) [inline include/geode/mesh/core/geode_regular_grid_solid.h:99]: I think we should remove these.

Comment (MelchiorSchuh) [reply include/geode/mesh/core/geode_regular_grid_solid.h]: I agree

File: include/geode/mesh/core/geode_regular_grid_solid.h [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+        const Point3D& origin() const override;
+
+        index_t cell_index( const GridCellIndices3D& index ) const override;
+
+        GridCellIndices3D cell_indices( index_t index ) const override;
+
+        index_t vertex_index( const GridVertexIndices3D& index ) const override;
+
+        GridVertexIndices3D vertex_indices( index_t index ) const override;
```
Comment (MelchiorSchuh) [inline include/geode/mesh/core/geode_regular_grid_solid.h:116]: Why are these private ? Don't we want to be able to access them ? Or is it temporary ?

Comment (BotellaA) [reply include/geode/mesh/core/geode_regular_grid_solid.h:116]: we don't want to access them directly, we want to call the virtual one and then use the vtable to call the correct implementation.

File: src/geode/mesh/builder/regular_grid_solid_builder.cpp [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        set_grid_dimensions(
+            std::move( cells_number ), std::move( cells_length ) );
+        grid_.vertex_attribute_manager().resize(
+            grid_.nb_vertices_in_direction( 0 )
+            * grid_.nb_vertices_in_direction( 1 )
+            * grid_.nb_vertices_in_direction( 2 ) );
+        grid_.polyhedron_attribute_manager().resize( grid_.nb_cells() );
+        for( const auto p : Range{ grid_.nb_polyhedra() } )
+        {
+            for( const auto v : LRange{ grid_.nb_polyhedron_vertices( p ) } )
```
Comment (panquez) [inline src/geode/mesh/builder/regular_grid_solid_builder.cpp:58]: LRange{ 8 } ?

File: src/geode/mesh/builder/regular_grid_surface_builder.cpp [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    {
+        set_grid_dimensions(
+            std::move( cells_number ), std::move( cells_length ) );
+        grid_.vertex_attribute_manager().resize(
+            grid_.nb_vertices_in_direction( 0 )
+            * grid_.nb_vertices_in_direction( 1 ) );
+        grid_.polygon_attribute_manager().resize( grid_.nb_cells() );
+        for( const auto p : Range{ grid_.nb_polygons() } )
+        {
+            for( const auto v : LRange{ grid_.nb_polygon_vertices( p ) } )
```
Comment (panquez) [inline src/geode/mesh/builder/regular_grid_surface_builder.cpp:59]: same question here?

File: src/geode/mesh/core/geode_regular_grid_solid.cpp [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+#include <geode/mesh/builder/regular_grid_solid_builder.h>
+#include <geode/mesh/core/private/grid_impl.h>
+#include <geode/mesh/core/private/points_impl.h>
+#include <geode/mesh/core/regular_grid_solid.h>
+
+namespace
+{
+    static constexpr std::array< std::array< geode::local_index_t, 3 >, 8 >
+        cell_vertices_translations{ { { 0, 0, 0 }, { 1, 0, 0 }, { 0, 1, 0 },
```
Comment (panquez) [inline src/geode/mesh/core/geode_regular_grid_solid.cpp:42]: translations? what does it mean?

File: src/geode/mesh/core/geode_regular_grid_surface.cpp [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+namespace
+{
+    static constexpr std::array< std::array< geode::local_index_t, 2 >, 4 >
+        cell_vertices_translations{ { { 0, 0 }, { 1, 0 }, { 0, 1 },
+            { 1, 1 } } };
+
+    // -X +X -Y +Y -Z +Z
+    static constexpr std::array< std::array< geode::local_index_t, 4 >, 6 >
+        cell_facet_vertices{ { { 0, 2, 6, 4 }, { 1, 5, 7, 2 }, { 0, 4, 5, 1 },
```
Comment (panquez) [inline src/geode/mesh/core/geode_regular_grid_surface.cpp:46]: facet for 2D grid? What does it represent?

File: src/geode/mesh/core/grid.cpp [hunk 7]
```diff
@@ -0,0 +1,541 @@
...
+
+        index_t nb_cells() const
+        {
+            index_t result{ 1 };
+            for( const auto d : LRange{ dimension } )
+            {
+                result *= nb_cells_in_direction( d );
+            }
+            return result;
+        }
```
Comment (panquez) [inline src/geode/mesh/core/grid.cpp:92]: duplication with l.74->78?

File: src/geode/mesh/core/geode_regular_grid_solid.cpp (hunk 5 above)
Comment (BotellaA) [reply src/geode/mesh/core/geode_regular_grid_solid.cpp:42]: When you have a CellIndices, it is how much you need to increment your indices to have the VertexIndices.

File: src/geode/mesh/builder/geode_regular_grid_solid_builder.cpp [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    void OpenGeodeRegularGridBuilder< 3 >::update_origin(
+        const Point3D& origin )
+    {
+        geode_regular_grid_.update_origin( origin, {} );
+    }
+
+    void OpenGeodeRegularGridBuilder< 3 >::do_create_vertices(
+        index_t /*unsued*/ )
```
Comment (MelchiorSchuh) [inline src/geode/mesh/builder/geode_regular_grid_solid_builder.cpp:53]: unused

File: include/geode/mesh/builder/geode_regular_grid_solid_builder.h [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    public:
+        OpenGeodeRegularGridBuilder(
+            VertexSet& vertex_set, MeshBuilderFactoryKey );
+
+        OpenGeodeRegularGridBuilder( OpenGeodeRegularGrid< 3 >& mesh );
+
+    private:
+        void update_origin( const Point3D& origin ) final;
+
+        void do_create_vertices( index_t /*unsued*/ ) final;
```
[the remark this answers is not in the document]
Comment (BotellaA) [reply include/geode/mesh/builder/geode_regular_grid_solid_builder.h:57]: the VertexSetBuilder::copy requires a do_create_vertices. So to remove this, we need to create and virtualize a `copy_vertices` methods that will prevent this call. Should we do that or leave it here?
@panquez

File: include/geode/mesh/helpers/private/regular_grid_shape_function.h [hunk 10]
```diff
@@ -24,11 +24,12 @@
 #pragma once
 
 #include <geode/mesh/common.h>
-#include <geode/mesh/core/regular_grid.h>
+#include <geode/mesh/core/grid.h>
 
 namespace geode
 {
     FORWARD_DECLARATION_DIMENSION_CLASS( Point );
+    FORWARD_DECLARATION_DIMENSION_CLASS( RegularGrid );
```
[the remark this answers is not in the document]
Comment (BotellaA) [reply include/geode/mesh/helpers/private/regular_grid_shape_function.h:32]: They are defined in grid.h which is included in this file.

**Pull request merged**